File: clients/elixir/lib/sharptown/transport/json_rpc.ex
defmodule Sharptown.Transport.JSONRPC do
  @moduledoc """
  The JSON-RPC transport — calls `image.transform` over a WebSocket at `{base_url}/rpc`.
  The image travels base64-encoded; the result is decoded back to raw bytes, so the returned
  `Sharptown.Response` matches the REST transport's shape.

  The scheme is optional: a bare `localhost:3002` resolves to `wss://localhost:3002`. Without
  a scheme the secure variant is used; pass `ws` (or `http`) explicitly to opt out.
  """

  @behaviour Sharptown.Transport

  alias Sharptown.{Error, JSON, Operations, Response}
  alias Sharptown.Net.WebSocket

  @impl true
  def transform(request, opts) do
    path = Keyword.get(opts, :path, "/rpc")
    method = Keyword.get(opts, :method, "image.transform")

    payload = %{
      "jsonrpc" => "2.0",
      "id" => 1,
      "method" => method,
      "params" => %{
        "image" => Base.encode64(request.bytes),
        "options" => Operations.to_options(request.operations)
      }
    }

    url = ws_endpoint(request.base_url, path)

    case WebSocket.connect(url, request.headers, request.timeout) do
      {:ok, conn} ->
        result = call(conn, JSON.encode(payload))
        WebSocket.close(conn)
        result

      {:error, %Error{} = error} ->
        {:error, error}
    end
  end

  defp call(conn, json) do
    with :ok <- WebSocket.send_text(conn, json),
         {:ok, raw} <- WebSocket.receive_text(conn) do
      decode(raw)
    end
  end

  defp decode(raw) do
    case JSON.decode(raw) do
      {:ok, %{"error" => error}} ->
        {:error, %Error{message: error_message(error), status: error_code(error), body: error}}

      {:ok, %{"result" => %{"image" => image} = result}} ->
        case Base.decode64(image) do
          {:ok, bytes} ->
            content_type = Map.get(result, "contentType", "application/octet-stream")
            {:ok, %Response{status: 200, headers: %{"content-type" => content_type}, body: bytes}}

          :error ->
            {:error, %Error{message: "JSON-RPC result.image is not valid base64"}}
        end

      _ ->
        {:error, %Error{message: "Malformed JSON-RPC response"}}
    end
  end

  defp error_message(%{"message" => message}) when is_binary(message), do: message
  defp error_message(_), do: "JSON-RPC error"

  defp error_code(%{"code" => code}) when is_integer(code), do: code
  defp error_code(_), do: nil

  defp ws_endpoint(base_url, path) do
    base = Sharptown.URL.ws_base(base_url)
    if has_path?(base), do: base, else: base <> path
  end

  defp has_path?(url) do
    case URI.parse(url).path do
      nil -> false
      "" -> false
      "/" -> false
      _ -> true
    end
  end
end

File: clients/flutter/lib/src/transport/json_rpc.dart
import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'package:web_socket_channel/web_socket_channel.dart';

import '../error.dart';
import '../operations.dart';
import '../response.dart';
import 'transport.dart';

/// The JSON-RPC over WebSocket transport. It calls `image.transform` at `{baseURL}/rpc`,
/// sending the image as base64 and receiving the result the same way.
class JsonRpcTransport implements Transport {
  /// Creates a JSON-RPC transport.
  const JsonRpcTransport({this.path = '/rpc', this.method = 'image.transform'});

  /// The WebSocket endpoint path.
  final String path;

  /// The JSON-RPC method name.
  final String method;

  @override
  Future<TransformResponse> transform(TransformRequest request) async {
    final resolved = await request.input.resolve(request.httpClient);

    final payload = jsonEncode({
      'jsonrpc': '2.0',
      'id': 1,
      'method': method,
      'params': {
        'image': base64Encode(resolved.bytes),
        'options': Operations.toOptions(request.operations),
      },
    });

    final channel =
        WebSocketChannel.connect(_wsEndpoint(request.baseUrl, path));
    try {
      await channel.ready.timeout(request.timeout);
    } catch (error) {
      throw SharptownError('WebSocket connection failed: $error');
    }

    channel.sink.add(payload);

    final Object? reply;
    try {
      reply = await channel.stream.first.timeout(request.timeout);
    } catch (error) {
      throw SharptownError('WebSocket read failed: $error');
    } finally {
      unawaited(channel.sink.close());
    }

    return _decode(reply);
  }

  static TransformResponse _decode(Object? reply) {
    final Map<String, Object?> message;
    try {
      message = jsonDecode(reply as String) as Map<String, Object?>;
    } catch (_) {
      throw const SharptownError('Malformed JSON-RPC response');
    }

    final error = message['error'];
    if (error is Map) {
      throw SharptownError(
        (error['message'] as String?) ?? 'JSON-RPC error',
        status: error['code'] as int?,
        body: error,
      );
    }

    final result = message['result'];
    final image = result is Map ? result['image'] : null;
    if (image is! String || image.isEmpty) {
      throw const SharptownError('JSON-RPC response is missing result.image');
    }

    final Uint8List bytes;
    try {
      bytes = base64Decode(image);
    } catch (_) {
      throw const SharptownError('JSON-RPC result.image is not valid base64');
    }

    final contentType =
        (result as Map)['contentType'] as String? ?? 'application/octet-stream';
    return TransformResponse(
      status: 200,
      headers: {'content-type': contentType},
      bytes: bytes,
    );
  }

  static Uri _wsEndpoint(String base, String path) {
    var normalized = base.trim();
    if (normalized.endsWith('/')) {
      normalized = normalized.substring(0, normalized.length - 1);
    }
    if (normalized.startsWith('http://')) {
      normalized = 'ws://${normalized.substring(7)}';
    } else if (normalized.startsWith('https://')) {
      normalized = 'wss://${normalized.substring(8)}';
    } else if (!normalized.startsWith('ws://') &&
        !normalized.startsWith('wss://')) {
      normalized = 'ws://$normalized';
    }

    final uri = Uri.parse(normalized);
    if (uri.path.isEmpty || uri.path == '/') {
      return uri.replace(path: path);
    }
    return uri;
  }
}
